### ISSUE TITLE
Julia interface to generated code

### ISSUE THREAD

**Issue by jaeandersson** [casadi/casadi#1730]
We now support generating both "mexFunction" and "main" entry points to to generated code, enabling the execution of the generated C code from MATLAB and the command line, respectively. There is also an open issue for generating a Python interface #1466.

It makes sense to add support for generating a Julia gateway to the generated code. This would allow the user to take an existing model created in CasADi C++, Python or MATLAB, generate code and e.g. prototype a solver algorithm in Julia.

I'm not very vested in how Julia interacts with external C code, but the MEX code can be generated and used as follows:
```matlab
import casadi.*

% Create some function
x = SX.sym('x', 5, 5);
F = Function('F', {x}, {gradient(det(x), x)});

% Generate C code with a MATLAB gateway
opts = struct('mex', true); % this would be replaced by "struct('julia', true);"
F.generate('my_codegen.c', opts);

% Compile the generated MEX function
mex my_codegen.c -largeArrayDims;

% Call the executed code
grad_det = my_codegen('F', rand(5)); % Note: the result is always a sparse matrix
disp(full(grad_det))
```

@mlubin @toivoh - do you have some tips on how this could be done?

**Comment by mlubin** [comment]
Julia just needs a shared library and a plain C signature to call.

**Comment by jaeandersson** [comment]
> Julia just needs a shared library and a plain C signature to call.

Well, you can do that from MATLAB as well, using "loadlibrary", but that's an API way too low level to be useful in practise.

So, essentially we would need to generate wrapper in Julia code which calls the shared library using `ccall`. A problem that remains though is how to compile/distribute this shared library in a cross-platform way. The generated code is already designed to be compiled into self-contained shared libraries and compiling is normally as easy as:
```
gcc -fPIC -shared my_casadi_codegen.c -o my_casadi_codegen.so
```
But this won't work on e.g. Windows. Does Julia have some standardized way to create shared libraries? Like MATLAB has the "mex" compiler command.

One interesting idea: We already distribute LLVM/Clang with the CasADi binaries for the purpose of just-in-time compilation. So we could actually pretty easily have CasADi export LLVM IR. Is there some way to have Julia use the LLVM IR instead of a shared library?

**Comment by jaeandersson** [comment]
It looks like there is an `llvmcall` command. http://docs.julialang.org/en/release-0.4/stdlib/c/
Maybe this could be used instead of `ccall`?

**Comment by mlubin** [comment]
Yep, ``llvmcall`` is definitely an option but I don't know anything about it!

**Comment by jaeandersson** [comment]
`

**Comment by jaeandersson** [comment]
That being said, `ccall` is probably much easier to get working.
